features/cellular/framework/AT/ATHandler.cpp

```diff
@@ -0,0 +1,1081 @@
+{
+    log_debug("%s", __func__);
+    if (consume_to_stop_tag()) {
+        set_scope(InfoType);
+    }
+}
+
+ATHandler::ScopeType ATHandler::get_scope()
+{
+    return _current_scope;
+}
+
+void ATHandler::set_string(char *dest, const char *src, size_t src_len)
+{
+    memcpy(dest, src, src_len);
+    dest[src_len] = '\0';
+}
+
+const char* ATHandler::mem_str(const char* dest, size_t dest_len, const char* src, size_t src_len)
+{
+   for(size_t i = 0; i < dest_len-src_len; ++i) {
+      if(memcmp(dest+i, src, src_len) == 0) {
+         return dest+i;
+      }
+   }
+   return NULL;
+}
+
+void ATHandler::cmd_start(const char* cmd)
+{
+    log_debug("AT> %s", cmd);
+
+    if (_last_err != NSAPI_ERROR_OK) {
+        return;
+    }
+
+    // write command
+    for (size_t i = 0; i < strlen(cmd); i++) {
+        if (write_char(cmd[i]) == false) {
+            // writing failed ---> write_char have set the last error, return...
+            return;
+        }
+    }
+
+    _cmd_start = true;
+}
+
+void ATHandler::write_int(int32_t param)
+{
+    // do common checks before sending subparameter
+    if (check_cmd_send() == false) {
+        return;
+    }
+
+    // write the integer subparameter
+    const uint8_t str_len = 12;
+    char number_string[str_len];
+    int32_t result = std::snprintf(number_string, str_len, "%ld", param);
+    if (result > 0 && result < str_len) {
+        for (size_t i = 0; number_string[i]; i++) {
+            if (write_char(number_string[i]) == false) {
+                // writing failed ---> write_char have set the last error, break out
+                break;
+            }
+        }
+    }
+}
+
+void ATHandler::write_string(const char* param, bool useQuotations)
+{
+    // do common checks before sending subparameter
+    if (check_cmd_send() == false) {
+        return;
+    }
+
+    // we are writing string, surround it with quotes
+    if (useQuotations && write_char('\"') == false) {
+        return;
+    }
+
+    for (size_t i = 0; i < strlen(param); i++) {
+        if (write_char(param[i]) == false) {
+            // writing failed ---> write_char have set the last error, return
+            break;
+        }
+    }
+
+    if (useQuotations) {
+        // we are writing string, surround it with quotes
+        write_char('\"');
+    }
+}
+
+void ATHandler::cmd_stop()
+{
+    if (_last_err != NSAPI_ERROR_OK) {
+        return;
+    }
+     // Finish with CR
+    for (size_t i = 0; i < _output_delimiter_length; i++) {
+        if (write_char(_output_delimiter[i]) == false) {
+            break;
+        }
+    }
+}
+
+size_t ATHandler::write_bytes(uint8_t *data, size_t len)
+{
+    if (_last_err != NSAPI_ERROR_OK) {
+        return 0;
+    }
+
+    size_t i = 0;
+    for (; i < len; i++) {
+        if (write_char(data[i]) == false) {
+            // writing failed ---> write_char have set the last error, return
+            break;
+        }
+    }
+
+    return i;
+}
+
+bool ATHandler::write_char(char c)
```

Comment:
Seems like it will also be quite important if using the mux as it stands - it seems that will spit out 1 mux frame per character if you send 1 character at a time.

Reply:
internal ticket created: IOTCELL-558